### PR TITLE
Nano: Add onnx inference support to nano

### PR DESCRIPTION
**Pull request by TheaperDeng**
issue #3174

### CODE REVIEW

File: python/nano/test/test_pytorch_onnx.py [hunk 1]
```diff
@@ -0,0 +1,110 @@
...
+from bigdl.nano.pytorch.trainer import Trainer
+from torch.optim import Adam
+from torch.utils.data import DataLoader, TensorDataset
+
+from bigdl.nano.pytorch.onnx import onnxruntime
+
+# adaptted from 
+# https://pytorch-lightning.readthedocs.io/en/latest/starter/introduction_guide.html
+# model definition
+@onnxruntime()
```
[the remark this answers is not in the document]
Comment (jason-dai) [reply python/nano/test/test_pytorch_onnx.py]: Will we be able to support keras in future?

[the remark this answers is not in the document]
Comment (TheaperDeng) [reply python/nano/test/test_pytorch_onnx.py]: I think so, but I am not sure if it can work in the same manner. https://github.com/onnx/tensorflow-onnx#python-api-reference
Maybe we can 
1. just put the decorator inside pytorch dir and move it out when keras support is ready. 
2. make it `nano.onnxruntime` and give some warning if users use it on a keras.Model now.

[the remark this answers is not in the document]
Comment (jason-dai) [reply python/nano/test/test_pytorch_onnx.py]: > I think so, but I am not sure if it can work in the same manner. [onnx/tensorflow-onnx#python-api-reference](https://github.com/onnx/tensorflow-onnx#python-api-reference) Maybe we can
> 
> 1. just put the decorator inside pytorch dir and move it out when keras support is ready.
> 2. make it `nano.onnxruntime` and give some warning if users use it on a keras.Model now.

Make it `nano.onnxruntime` and report error ("tf not currently supported") for now?

[the remark this answers is not in the document]
Comment (TheaperDeng) [reply python/nano/test/test_pytorch_onnx.py]: currently I will just implement `trainer.compile(onnx=True)` and make all our users with `nn.module` to use this API. We believe that is the unified API that can handle all the use cases.

[the remark this answers is not in the document]
Comment (jason-dai) [reply python/nano/test/test_pytorch_onnx.py]: Should be `trainer.compile(model, onnx=True)`? Or `nano.pytorch.Trainer.compile(model, onnx=True)`?

[the remark this answers is not in the document]
Comment (TheaperDeng) [reply python/nano/test/test_pytorch_onnx.py]: > Should be `trainer.compile(model, onnx=True)`? Or `nano.pytorch.Trainer.compile(model, onnx=True)`?

Since trainer is not binded to any model instance, there is not much of a difference between these two. a `trainer.compile(model, onnx=True)` is good enough I think. of course, trainer is an instance of `nano.pytorch.Trainer`

[the remark this answers is not in the document]
Comment (jason-dai) [reply python/nano/test/test_pytorch_onnx.py]: I think the only difference is if the user needs to first create a `Trainer` instance or not; and `trainer.compile` does not bind `model` to `trainer`, which may be a little bit odd?

[the remark this answers is not in the document]
Comment (TheaperDeng) [reply python/nano/test/test_pytorch_onnx.py]: > I think the only difference is if the user needs to first create a `Trainer` instance or not; and `trainer.compile` does not bind `model` to `trainer`, which may be a little bit odd?

Well, since they need trainer to `fit` or `predict` as long as they want to use our acceleration, I think we are good. btw, `trainer.fit` and other methods in trainer does not bind model to trainer as well.

[the remark this answers is not in the document]
Comment (jason-dai) [reply python/nano/test/test_pytorch_onnx.py]: ok

[the remark this answers is not in the document]
Comment (TheaperDeng) [reply python/nano/test/test_pytorch_onnx.py]: oh, I just though about a thing. What if a user has a pl.lightningmodule and he/she just want to use onnx inference support...? It would be odd to use `trainer.compile` to support this case...

[the remark this answers is not in the document]
Comment (jason-dai) [reply python/nano/test/test_pytorch_onnx.py]: I think decorator alone is problematic if the class is not defined by the user (e.g., imported from a 3rd party library); you need to use some method to convert either the class or the instance. Not sure if you want to have another method (say `onnxrt`) or just use `compile`

[the remark this answers is not in the document]
Comment (TheaperDeng) [reply python/nano/test/test_pytorch_onnx.py]: > I think decorator alone is problematic if the class is not defined by the user (e.g., imported from a 3rd party library); you need to use some method to convert either the class or the instance. Not sure if you want to have another method (say `onnxrt`) or just use `compile`

The `trainer.compile` thing can definitely handle all the use case when user only have a `nn.module` class, instance or create function. This API will be implemented after the pl module PR.

For the user who have a pl.lightningmodule, I will find a way to support it. My current plan is `trainer.compile(model, onnx=True)`, when the function found that `model` is a pl.lightningmodule, it will ignore the `loss` and `opt` param and just add the onnx support.

The decorator thing will be eliminated.

[the remark this answers is not in the document]
Comment (jason-dai) [reply python/nano/test/test_pytorch_onnx.py]: > > I think the only difference is if the user needs to first create a `Trainer` instance or not; and `trainer.compile` does not bind `model` to `trainer`, which may be a little bit odd?
> 
> Well, since they need trainer to `fit` or `predict` as long as they want to use our acceleration, I think we are good. btw, `trainer.fit` and other methods in trainer does not bind model to trainer as well.

One question - if the model is defined in one library (e.g., Chornos) but used in another place, should the one that defines the model to call `compile`, or the one that uses the model to call `compile`? If it's the former, will it be odd that he or she must first define a trainer, or you think it's OK?

[the remark this answers is not in the document]
Comment (TheaperDeng) [reply python/nano/test/test_pytorch_onnx.py]: > 
> 
> > > I think the only difference is if the user needs to first create a `Trainer` instance or not; and `trainer.compile` does not bind `model` to `trainer`, which may be a little bit odd?
> > 
> > 
> > Well, since they need trainer to `fit` or `predict` as long as they want to use our acceleration, I think we are good. btw, `trainer.fit` and other methods in trainer does not bind model to trainer as well.
> 
> One question - if the model is defined in one library (e.g., Chornos) but used in another place, should the one that defines the model to call `compile`, or the one that uses the model to call `compile`? If it's the former, will it be odd that he or she must first define a trainer, or you think it's OK?

**Of course the one who uses the model should call `compile`.** ONNXRuntime support is only an extension on a **model instance**, and the one who defines the model only give the network topology, forwarding (and training loop if valid). The  users who want to use that model will instantiate the model and call `compile` to add the ONNXRuntime support to its model instance.

**Comment by TheaperDeng** [comment]
@yangw1234 

1. Since onnx support is directly binded to a **model instance**, it would be hard to follow and pass typecheck. I close it in `mypy.ini`.
2. Should we defaultly install onnx and onnxruntime in `bigdl-nano[pytorch]`?

**Comment by yangw1234** [comment]
> @yangw1234
> 
> 1. Since onnx support is directly binded to a **model instance**, it would be hard to follow and pass typecheck. I close the test for onnx in `mypy.ini`.
> 2. Should we defaultly install onnx and onnxruntime in `bigdl-nano[pytorch]`?

Sounds good to me

**Comment by TheaperDeng** [comment]
http://10.112.231.51:18888/job/ZOO-PR-NanoTests/91/

File: python/nano/src/bigdl/nano/pytorch/trainer/Trainer.py [hunk 2]
```diff
@@ -122,15 +127,29 @@ def compile(model: nn.Module, loss: _Loss = None, optimizer: torch.optim.Optimiz
...
                 "Loss and optimizer are required to construct a LightningModule instance."
-            return LightningModuleFromTorch(model, loss, optimizer)
+            pl_model = LightningModuleFromTorch(model, loss, optimizer)
+
+        if onnx:
+            try:
+                from bigdl.nano.pytorch.onnx.onnxrt_inference import bind_onnxrt_methods
+                return bind_onnxrt_methods(pl_model)
+            except ImportError:
+                warnings.warn("You should install onnx and onnxruntime to set `onnx=True`")
```
Comment (yangw1234) [inline python/nano/src/bigdl/nano/pytorch/trainer/Trainer.py:152]: Maybe we should raise an error instead of just a warning, because user is explictly saying he/she wants to use `onnx`, skipping binding is not the intended behavior. He/She might not even notice the warning.

(Or send a clearer and eye-catching message, such as 
```
############################################################################
############################################################################
##################         Warning    ######################################
The argument `onnx` is set to True, but onnx and onnxruntime are installed. 
The argument is ignored. To use `onnx`, please install onnx and onnxruntime.
############################################################################
############################################################################
```

File: python/nano/src/bigdl/nano/pytorch/onnx/onnxrt_inference.py [hunk 3]
```diff
@@ -0,0 +1,199 @@
...
+
+def bind_onnxrt_methods(pl_model: LightningModule):
+    # class type check
+    assert isinstance(pl_model, LightningModule),\
+        "onnxruntime support is only valid for a LightningModule."
+
+    # check conflicts
+    for component in ONNXRT_BINDED_COMPONENTS:
+        if component in dir(pl_model):
+            warnings.warn(f"{component} method/property will be replaced.")
```
Comment (yangw1234) [inline python/nano/src/bigdl/nano/pytorch/onnx/onnxrt_inference.py:42]: Maybe explaining why this happened and what users can do to make this warning disappear. If there is nothing user can do, maybe it is better to use `logging.warn`.

image


see https://docs.python.org/3/howto/logging.html

File: python/nano/src/bigdl/nano/pytorch/onnx/onnxrt_inference.py [hunk 4]
```diff
@@ -0,0 +1,199 @@
...
+        :param sess_options: ortsess options in ort.SessionOptions type.
+        :param backend: str, to set the backend library. "onnx" for onnxruntime, which
+               provides lower latency and any other value will make `inference` call
+               the pytorch forwarding method.
+        :param **kwargs: any other keywords that will be passed to onnx session's building.
+        '''
+
+        if backend == "onnx":
+            if not self._ortsess_up_to_date:
+                warnings.warn("Onnxruntime session will be built implicitly,"
```
Comment (yangw1234) [inline python/nano/src/bigdl/nano/pytorch/onnx/onnxrt_inference.py:158]: Maybe also log the building time.

File: python/nano/src/bigdl/nano/pytorch/onnx/onnxrt_inference.py [hunk 5]
```diff
@@ -0,0 +1,199 @@
...
+                # this branch is only to speed up the inferencing when batch_size is set to None.
+                ort_inputs = {input_name: input_data}
+                ort_outs = self._ortsess.run(None, ort_inputs)
+                return ort_outs[0]
+            else:
+                yhat_list = []
+                sample_num = input_data.shape[0]  # the first dim should be sample_num
+                batch_num = math.ceil(sample_num / batch_size)
+                for batch_id in range(batch_num):
+                    ort_inputs = {input_name: input_data[batch_id * batch_size:
```
Comment (yangw1234) [inline python/nano/src/bigdl/nano/pytorch/onnx/onnxrt_inference.py:176]: Does not support multiple inputs?

**Comment by TheaperDeng** [comment]
I will raise another PR to fix according to the comments later.

File: python/nano/src/bigdl/nano/pytorch/onnx/onnxrt_inference.py (hunk 5 above)
Comment (TheaperDeng) [reply python/nano/src/bigdl/nano/pytorch/onnx/onnxrt_inference.py:176]: yes

File: python/nano/src/bigdl/nano/pytorch/onnx/onnxrt_inference.py [hunk 6]
```diff
@@ -0,0 +1,199 @@
...
+                                    'dynamic_axes': {'input': {0: 'batch_size'},
+                                                     'output': {0: 'batch_size'}}}
+        default_onnx_export_args.update(kwargs)
+
+        self.to_onnx(file_path,
+                     input_sample,
+                     **default_onnx_export_args)
+
+        self._ortsess = ort.InferenceSession(file_path, sess_options=sess_options)
+        self._ortsess_up_to_date = True
```
Comment (jason-dai) [inline python/nano/src/bigdl/nano/pytorch/onnx/onnxrt_inference.py:85]: how do we clean up and release the resources?

File: python/nano/src/bigdl/nano/pytorch/onnx/onnxrt_inference.py [hunk 7]
```diff
@@ -0,0 +1,199 @@
...
+                            file_path=file_path,
+                            sess_options=sess_options,
+                            **kwargs)
+
+    pl_model.update_ortsess = partial(update_ortsess, pl_model)
+
+    # on_fit_start (LightningModule method override)
+    def on_fit_start_additional(function):
+        def wrapped(*args, **kwargs):
+            args[0]._ortsess_up_to_date = False
```
Comment (jason-dai) [inline python/nano/src/bigdl/nano/pytorch/onnx/onnxrt_inference.py:115]: shall we also clean up and release the resources here?

File: python/nano/src/bigdl/nano/pytorch/onnx/onnxrt_inference.py [hunk 8]
```diff
@@ -0,0 +1,199 @@
...
+        else:
+            # inference w/o onnxruntime (fallback to pytorch native forward)
+            self.eval()
+            with torch.no_grad():
+                yhat_list = []
+                sample_num = input_data.shape[0]  # the first dim should be sample_num
+                batch_size = batch_size if batch_size else sample_num
+                batch_num = math.ceil(sample_num / batch_size)
+                for batch_id in range(batch_num):
+                    yhat_list.append(self(input_data[batch_id * batch_size:
```
Comment (jason-dai) [inline python/nano/src/bigdl/nano/pytorch/onnx/onnxrt_inference.py:192]: shall we call forward or predict_step here?

File: python/nano/src/bigdl/nano/pytorch/onnx/onnxrt_inference.py (hunk 7 above)
Comment (TheaperDeng) [reply python/nano/src/bigdl/nano/pytorch/onnx/onnxrt_inference.py:115]: good idea, I will release the resource here

File: python/nano/src/bigdl/nano/pytorch/onnx/onnxrt_inference.py (hunk 8 above)
Comment (TheaperDeng) [reply python/nano/src/bigdl/nano/pytorch/onnx/onnxrt_inference.py:192]: I won't use `predict_step` here because `predict_step`  is not always defined in users code. So I use `self()` to call `forward` here.
